runtime/doc/msgpack_rpc.txt

```diff
@@ -404,4 +404,91 @@ of update.
 	The popupmenu is hidden.
 
 ==============================================================================
+7. Watching buffers						*rpc--watching-buffers*
+
+Special functions have been created to allow co-processes to be notified in
+real-time when a buffer's content changes. (It is difficult and innefficient
+to do this using |TextChanged| and other autocommands.)
+
+		*Todo	what about when a file is changed from outside vim?
+				(See |FileChangedShell|, |ShellCmdPost| and |ShellFilterPost|)
+
+Setting Up~
+
+Use |rpcstart()| to start up your co-process. This will give you a valid
+{channel} to work with. >
+  let g:mychannel = rpcstart(...)
+
+Next, use |rpcbufwatch()| to tell nvim to start sending notifications to your
+{channel}. The simplest way to get a {bufnr} to work with is by using the
+|bufnr()| function. For example, you can use `bufnr("")` to get the {bufnr} of
+the current buffer.
+  call rpcbufwatch(bufnr(""), g:mychannel)
+
+Receiving Events~
+
+The co-process will start receiving three different types of notification
+events which will be equivilent to the following |rpcnotify()| calls:
+
+1. rpcnotify({channel}, 'BufLinesAdded', [{int1}, {int2}])
+
+	Indicates that one or more new lines were added to the buffer.  The new
+	lines were added after the line number in {int1}. If {int1} is `0`, then
+	the new lines were added at the top of the file. {int1} Will never be
+	greater than the last line number.  {int2} tells how many lines were added
+	and will always be greater than `0`.
+
+2. rpcnotify({channel}, 'BufLinesRemoved', [{int1}, {int2}])
+
+	Indicates that one or more lines were removed from the buffer.  {int1} is
+	the line number of the first line that was removed and will always be
+	between `1` and `line('$')`. {int2} tells how many lines were removed and
+	will always be at least `1`.
+
+	*Todo	could just use BufLinesReplaced with int3 = 0 for this
+
+3. rpcnotify({channel}, 'BufLinesReplaced', [{int1}, {int2}, {int3}])
+
+	Indicates that one or more lines from the buffer have been removed and
+	replaced with new lines. {int1} is the line number of the first line that
+	was replaced and will always be between `1` and `line('$')`. {int2} tells
+	how many lines were replaced and will always be at least `1`. For most
+	small changes {int2} will be `1`, meaning only the line mentioned in {int1}
+	was changed. {int3} tells how many new lines are replacing the lines that
+	were removed.
+
+	For example, if the user selects lines 3-5 of a buffer using |Visual-mode|
+	and then presses `p` to paste in new block of 6 lines, then the following
+	event would be sent to the co-process: >
+		rpcnotify({channel}, 'BufLinesReplaced', [3, 3, 6])
```

Comment:
I think it should use  zero-based indexing just like the buffer api methods like `nvim_buf_set_lines`, so this would be `[2, 5, 6]`. Then `BufLinesAdded` (and `BufLinesRemoved` as you noted) would just be special cases of this, i e `[1,1,3]` to add three lines after line 1.